client-ts/Microsoft.AspNetCore.SignalR.Client.TS/Formatters.ts

```diff
@@ -21,55 +21,71 @@ export namespace TextMessageFormat {
 }
 
 export namespace BinaryMessageFormat {
+
+    // The length prefix of binary messages is encoded as VarInt. Read the comment in
+    // the BinaryMessageParser.TryParseMessage for details.
+
     export function write(output: Uint8Array): ArrayBuffer {
         // .byteLength does is undefined in IE10
         let size = output.byteLength || output.length;
-        let buffer = new Uint8Array(size + 8);
-
-        // javascript bitwise operators only support 32-bit integers
-        for (let i = 7; i >= 4; i--) {
-            buffer[i] = size & 0xff;
-            size = size >> 8;
+        let lenBuffer = [];
+        do
+        {
+            let sizePart = size & 0x7f;
+            size = size >> 7;
+            if (size > 0) {
+                sizePart |= 0x80;
+            }
+            lenBuffer.push(sizePart);
         }
+        while (size > 0);
 
-        buffer.set(output, 8);
+        // .byteLength does is undefined in IE10
+        size = output.byteLength || output.length;
 
+        let buffer = new Uint8Array(lenBuffer.length + size);
+        buffer.set(lenBuffer, 0);
+        buffer.set(output, lenBuffer.length);
         return buffer.buffer;
     }
 
     export function parse(input: ArrayBuffer): Uint8Array[] {
         let result: Uint8Array[] = [];
         let uint8Array = new Uint8Array(input);
-        // 8 - the length prefix size
-        for (let offset = 0; offset < input.byteLength;) {
+        const MaxLengthPrefixSize = 5;
+        const NumBitsToShift = [0, 7, 14, 21, 28 ];
 
-            if (input.byteLength < offset + 8) {
-                throw new Error("Cannot read message size")
+        for (let offset = 0; offset < input.byteLength;) {
```

Comment:
Why doesn't this also look at byteLength and length?

Reply:
Need to double check - most likely it should.

Reply:
I was baffled by this because the tests did not fail. Now I know why. For IE10 msgpack returns Buffer and not UInt8Array and Buffer does not have byteLength. We create an Uint8Array directly which does have byteLength.  They have a special flag TYPED_ARRAY_SUPPORT they use to decide what to return and the comment says:

/**
 * If `Buffer.TYPED_ARRAY_SUPPORT`:
 *   === true    Use Uint8Array implementation (fastest)
 *   === false   Use Object implementation (most compatible, even IE6)
 *
 * Browsers that support typed arrays are IE 10+, Firefox 4+, Chrome 7+, Safari 5.1+,
 * Opera 11.6+, iOS 4.2+.
 *
 * Due to various browser bugs, sometimes the Object implementation will be used even
 * when the browser supports typed arrays.
 *
 * Note:
 *
 *   - Firefox 4-29 lacks support for adding new properties to `Uint8Array` instances,
 *     See: https://bugzilla.mozilla.org/show_bug.cgi?id=695438.
 *
 *   - Chrome 9-10 is missing the `TypedArray.prototype.subarray` function.
 *
 *   - IE10 has a broken `TypedArray.prototype.subarray` function which returns arrays of
 *     incorrect length in some situations.

 * We detect these buggy browsers and set `Buffer.TYPED_ARRAY_SUPPORT` to `false` so they
 * get the Object implementation, which is slower but behaves correctly.
 */

We do use subarray in one place because IE does not support to slice so I need to dig a bit and understand how subarray is broken in IE10.